File: Trees_Graphs/go/sequences.go
package main
import (
    "fmt"
    "types"
)


func getSequences(node *types.TreeNode) [][]*types.TreeNode {
  var sequence []*types.TreeNode
  var listOfSequences, leftSequences, rightSequences [][]*types.TreeNode

  if node.LeftChild == nil && node.RightChild == nil {
    sequence = append(sequence, node)
    listOfSequences = append(listOfSequences, sequence)
  } else if node.LeftChild == nil {
    rightSequences = getSequences(node.RightChild)
    for i := 0; i < len(rightSequences); i++ {
      var s []*types.TreeNode
      s = append(s, node)
      sequence = rightSequences[i]

      s = append(s, sequence...)

      listOfSequences = append(listOfSequences, s)
    }
  } else if node.RightChild == nil {
    leftSequences = getSequences(node.LeftChild)
    for i := 0; i < len(leftSequences); i++ {
      var s []*types.TreeNode
      s = append(s, node)
      sequence = leftSequences[i]

      s = append(s, sequence...)

      listOfSequences = append(listOfSequences, s)
    }
  } else {
    leftSequences = getSequences(node.LeftChild)
    rightSequences = getSequences(node.RightChild)

    // fmt.Println("left")
    // for i := 0; i < len(leftSequences); i++ {
    //   for j := 0; j < len(leftSequences[i]); j++ {
    //     fmt.Printf("%v ", leftSequences[i][j].Value)
    //   }
    //   fmt.Println("\n")
    // }
    // fmt.Println("right")
    // for i := 0; i < len(rightSequences); i++ {
    //   for j := 0; j < len(rightSequences[i]); j++ {
    //     fmt.Printf("%v ", rightSequences[i][j].Value)
    //   }
    //   fmt.Println("\n")
    // }

    for i := 0; i < len(leftSequences); i++ {
      var begin []*types.TreeNode
      sequence = leftSequences[i]
      begin = append(begin, node)
      begin = append(begin, sequence...)

      for j := 0; j < len(rightSequences); j++ {
        var s []*types.TreeNode
        sequence = rightSequences[j]
        s = append(s, begin...)
        s = append(s, sequence...)
        listOfSequences = append(listOfSequences, s)
      }
    }

    for i := 0; i < len(rightSequences); i++ {
      var begin []*types.TreeNode
      sequence = rightSequences[i]
      begin = append(begin, node)
      begin = append(begin, sequence...)

      for j := 0; j < len(leftSequences); j++ {
        var s []*types.TreeNode
        sequence = leftSequences[j]
        s = append(s, begin...)
        s = append(s, sequence...)
        listOfSequences = append(listOfSequences, s)
      }
    }
  }

  return listOfSequences
}


func main() {
  array := []int{1, 2, 3, 4, 5, 6, 7}
  root := types.CreateTree(array)
  types.PrintTreeNodes(root)

  sequences := getSequences(root)

  for i := 0; i < len(sequences); i++ {
    for j := 0; j < len(sequences[i]); j++ {
      fmt.Printf("%v ", sequences[i][j].Value)
    }
    fmt.Println("\n")
  }
}

File: Trees_Graphs/python/sequences.py
from Tree import *


def get_sequences(node):
    sequences, list_of_sequences = [], []

    if node.left_child is None and node.right_child is None:
        sequences.append(node)
        list_of_sequences.append(sequences)
    elif node.left_child is None:
        right_sequences = get_sequences(node.right_child)

        for sequence in right_sequences:
            s = [node, *sequence]
            list_of_sequences.append(s)
    elif node.right_child is None:
        left_sequences = get_sequences(node.left_child)

        for sequence in left_sequences:
            s = [node, *sequence]
            list_of_sequences.append(s)
    else:
        left_sequences = get_sequences(node.left_child)
        right_sequences = get_sequences(node.right_child)

        for sequence in left_sequences:
            begin = [node, *sequence]

            for sequence in right_sequences:
                s = [*begin, *sequence]
                list_of_sequences.append(s)

        for sequence in right_sequences:
            begin = [node, *sequence]

            for sequence in left_sequences:
                s = [*begin, *sequence]
                list_of_sequences.append(s)

    return list_of_sequences

def main():
    array = list(range(7))
    tree = Tree(array)
    tree.print_tree()

    print(get_sequences(tree.root))


if __name__ == '__main__':
    main()
